src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp

```diff
@@ -46,18 +47,91 @@ depends_on(const exprt &expr, std::unordered_set<irep_idt> identifiers)
   return false;
 }
 
+/// Collect identifiers that are local to this loop only
+/// (excluding nested loop).
+/// A target is a loop local if
+/// 1. it is declared inside the loop, or
+/// 2. there is no write or read of it outside the loop.
+std::unordered_set<irep_idt> gen_loop_locals_set(
+  const irep_idt &function_id,
+  goto_functiont &goto_function,
+  const loopt &loop,
+  message_handlert &message_handler,
+  const namespacet &ns)
+{
+  std::unordered_set<irep_idt> loop_locals;
+  std::unordered_set<irep_idt> non_loop_locals;
+  // Ranges of all variables declared outside the loop.
+  rw_range_sett decl_rw_range_set(ns, message_handler);
+  // Ranges of all read/write outside the loop.
+  rw_range_sett non_loop_rw_range_set(ns, message_handler);
+
+  Forall_goto_program_instructions(i_it, goto_function.body)
+  {
+    // All variables declared in loops are loop locals.
+    if(i_it->is_decl() && loop.contains(i_it))
+    {
+      loop_locals.insert(i_it->decl_symbol().get_identifier());
+    }
+    // Record all other declared variables and their ranges.
+    else if(i_it->is_decl())
+    {
+      goto_rw(function_id, i_it, decl_rw_range_set);
```

Comment:
Ah, I think this explains why you had to add those `nondet_int()` calls. See comment below.

Reply:
See comment above.